### PR TITLE
use google-double conversion

### LINKED ISSUE #7: Use double-conversion to convert PyFloat
* Amalgamate double-conversion
* Export C interface for converting float and use in Py-CJSON

### PR DESCRIPTION
**Pull request by Nambers**
TODO
- [x] free in some points
- [x] validate string in decoding
- [x] ~ensure to throw overflow exception when double is invalid? align to std python behavior~ just act like inf

Closed #7

### CODE REVIEW

File: buildScripts/amalgamation/amalgamation.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+import json
```
Comment (Antares0982) [inline buildScripts/amalgamation/amalgamation.py:1]: remove these files

File: deps/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+double-conversion/**
```
Comment (Antares0982) [inline deps/.gitignore:1]: remove this file

File: src/pycJSON_decode.c [hunk 3]
```diff
@@ -490,21 +491,34 @@ static bool parse_number(PyObject **item, parse_buffer *const input_buffer) {
     }
 
     if (dec) {
-        const double temp = PyOS_string_to_double((const char *) starting_point, (char **) &after_end, PyExc_OverflowError);
-        if (PyErr_Occurred()) return false;
+        // const double temp = PyOS_string_to_double((const char *) starting_point, (char **) &after_end, PyExc_OverflowError);
+        // if (PyErr_Occurred()) return false;
+        if (dconv_s2d_ptr == NULL) {
```
Comment (Antares0982) [inline src/pycJSON_decode.c:496]: init in PyInit_cjson

**Review by Antares0982 — CHANGES_REQUESTED** [review]

File: src/pycJSON_decode.c (hunk 3 above)
Comment (Nambers) [reply src/pycJSON_decode.c:496]: 驳回!

File: src/dconv_wrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#include "dconv_wrapper.h"
+#include "double-conversion.hpp"
+
+#include <Python.h>
+#include <pymath.h>
+
+namespace double_conversion {
+    extern "C" {
+    int dconv_d2s(double value, char *buf, int buflen, int *strlength) {
+        static auto d2s = DoubleToStringConverter(DoubleToStringConverter::NO_FLAGS, "Infinity", "NaN", 'e', -324, 308, 0, 0);
```
Comment (Antares0982) [inline src/dconv_wrapper.cpp:10]: static DoubleToStringConverter d2s(DoubleToStringConverter::NO_FLAGS, "Infinity", "NaN", 'e', -324, 308, 0, 0);
```suggestion
        static DoubleToStringConverter d2s(DoubleToStringConverter::NO_FLAGS, "Infinity", "NaN", 'e', -324, 308, 0, 0);
```
This fixes compile error

File: src/dconv_wrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        StringBuilder sb(buf, buflen);
+        int success = d2s.ToShortest(value, &sb);
+        *strlength = success ? sb.position() : -1;
+        return success;
+    }
+
+
+    double dconv_s2d(const char *buffer, int length, int *processed_characters_count) {
+        static auto s2d = StringToDoubleConverter(StringToDoubleConverter::NO_FLAGS, 0.0, Py_NAN, "Infinity", "NaN");
```
Comment (Antares0982) [inline src/dconv_wrapper.cpp:19]:
```suggestion
        static StringToDoubleConverter s2d(StringToDoubleConverter::NO_FLAGS, 0.0, Py_NAN, "Infinity", "NaN");
```

**Review by Antares0982 — CHANGES_REQUESTED** [review]

**Pull request merged**